### PR TITLE
 [SYCL-JIT] Store env variables in configuration

### PR DESCRIPTION
**Pull request by jchlanda**
By storing the env variables in the config helper we eliminate the need
for extra function parameters while also making it accessible throughout
the codebase.

### CODE REVIEW

File: sycl-jit/jit-compiler/include/Options.h [hunk 1]
```diff
@@ -90,6 +96,17 @@ struct JITTargetInfo
...
+struct JITTargetCPU
+    : public OptionBase<JITTargetCPU, OptionID::TargetCPU, std::string> {
+  using OptionBase::OptionBase;
+};
+
+struct JITTargetFeatures
+    : public OptionBase<JITTargetFeatures, OptionID::TargetFeatures,
+                        std::string> {
+  using OptionBase::OptionBase;
+};
```
Comment (victor-eds) [inline sycl-jit/jit-compiler/include/Options.h:108]: I'm not sure this `std::string` usage is safe here (linking SYCL and SYCL-JIT with different STL lib versions). Can we use the `DynArray<char>` instead?

Comment (jchlanda) [reply sycl-jit/jit-compiler/include/Options.h]: Good shout, [done](https://github.com/intel/llvm/pull/14860/commits/86f92194f8a64bf06e8782b057138e9944780e7c).

File: sycl-jit/jit-compiler/lib/translation/KernelTranslation.cpp [hunk 2]
```diff
@@ -261,8 +260,9 @@ llvm::Expected<KernelBinary *> KernelTranslator::translateToPTX(
 
   // Give priority to user specified values (through environment variables:
   // SYCL_JIT_AMDGCN_PTX_TARGET_CPU and SYCL_JIT_AMDGCN_PTX_TARGET_FEATURES).
-  llvm::StringRef CPU{TargetCPU};
-  llvm::StringRef Features{TargetFeatures};
+  llvm::StringRef CPU = ConfigHelper::get<option::JITTargetCPU>().begin();
+  llvm::StringRef Features =
+      ConfigHelper::get<option::JITTargetFeatures>().begin();
```
Comment (victor-eds) [inline sycl-jit/jit-compiler/lib/translation/KernelTranslation.cpp:265]: `ConfigHelper::get` returns a temporary, so `CPU` and `Features` contain dangling pointers, right?

Comment (jchlanda) [reply sycl-jit/jit-compiler/lib/translation/KernelTranslation.cpp:265]: Yeap, this needs a bit more work, my build has tricked me again :) will ping you later.

Comment (victor-eds) [reply sycl-jit/jit-compiler/lib/translation/KernelTranslation.cpp:265]: Thanks!

Comment (jchlanda) [reply sycl-jit/jit-compiler/lib/translation/KernelTranslation.cpp]: I've given them a concrete storage of a `string` and fixed `DynArray` creation, should be OK now @victor-eds.

File: sycl-jit/jit-compiler/lib/translation/KernelTranslation.cpp [hunk 3]
```diff
@@ -260,9 +260,8 @@ KernelTranslator::translateToPTX(SYCLKernelInfo &KernelInfo, llvm::Module &Mod,
 
   // Give priority to user specified values (through environment variables:
   // SYCL_JIT_AMDGCN_PTX_TARGET_CPU and SYCL_JIT_AMDGCN_PTX_TARGET_FEATURES).
-  llvm::StringRef CPU = ConfigHelper::get<option::JITTargetCPU>().begin();
-  llvm::StringRef Features =
-      ConfigHelper::get<option::JITTargetFeatures>().begin();
+  std::string CPU = ConfigHelper::get<option::JITTargetCPU>().begin();
+  std::string Features = ConfigHelper::get<option::JITTargetFeatures>().begin();
```
Comment (victor-eds) [inline sycl-jit/jit-compiler/lib/translation/KernelTranslation.cpp:264]:
```suggestion
  auto CPUVal = ConfigHelper::get<option::JITTargetCPU>();
  auto FeaturesVal = ConfigHelper::get<option::JITTargetFeatures>();
  llvm::StringRef CPU = CPUVal.begin();
  llvm::StringRef Features = FeaturesVal.begin();
```

Can we do something like this instead? (and below)

File: sycl/source/detail/jit_compiler.cpp [hunk 4]
```diff
@@ -664,14 +664,22 @@ ur_kernel_handle_t jit_compiler::materializeSpecConstants(
...
 
-  std::string TargetCPU =
-      detail::SYCLConfig<detail::SYCL_JIT_AMDGCN_PTX_TARGET_CPU>::get();
-  std::string TargetFeatures =
-      detail::SYCLConfig<detail::SYCL_JIT_AMDGCN_PTX_TARGET_FEATURES>::get();
+  auto SetUpOption = [](std::string Value) {
+    ::jit_compiler::JITEnvVar Option(Value.size());
+    memcpy(static_cast<void *>(Option.begin()),
+           static_cast<const void *>(Value.data()), Value.size());
+    return Option;
```
Comment (victor-eds) [inline sycl/source/detail/jit_compiler.cpp:671]:
```suggestion
    return ::jit_compiler::JITEnvVar{Value.size(), Value.data()};
```

And implement constructor in `DynArray`

File: sycl/source/detail/jit_compiler.cpp [hunk 5]
```diff
@@ -664,14 +664,22 @@ ur_kernel_handle_t jit_compiler::materializeSpecConstants(
   AddToConfigHandle(
       ::jit_compiler::option::JITEnableVerbose::set(DebugEnabled));
 
-  std::string TargetCPU =
-      detail::SYCLConfig<detail::SYCL_JIT_AMDGCN_PTX_TARGET_CPU>::get();
-  std::string TargetFeatures =
-      detail::SYCLConfig<detail::SYCL_JIT_AMDGCN_PTX_TARGET_FEATURES>::get();
+  auto SetUpOption = [](std::string Value) {
```
Comment (victor-eds) [inline sycl/source/detail/jit_compiler.cpp:667]:
```suggestion
  auto SetUpOption = [](const std::string &Value) {
```

File: sycl-jit/jit-compiler/lib/translation/KernelTranslation.cpp (hunk 3 above)
Comment (jchlanda) [reply sycl-jit/jit-compiler/lib/translation/KernelTranslation.cpp]: Yeap, done.

File: sycl-jit/common/include/DynArray.h [hunk 6]
```diff
@@ -22,6 +23,11 @@ template <typename T> class DynArray {
 
   explicit DynArray(size_t Size) { init(Size); }
 
+  explicit DynArray(size_t Size, const void *Data) {
+    init(Size);
+    memcpy(static_cast<void *>(this->begin()), Data, Size * sizeof(T));
+  }
```
Comment (victor-eds) [inline sycl-jit/common/include/DynArray.h:29]:
```suggestion
  template <typename InputIt>
  DynArray(InputIt begin, InputIt end) {
    init(end - begin);
    std::copy(begin, end, this->begin());
  }
```

What about this?

File: sycl/source/detail/jit_compiler.cpp [hunk 7]
```diff
@@ -663,11 +663,8 @@ ur_kernel_handle_t jit_compiler::materializeSpecConstants(
       detail::SYCLConfig<detail::SYCL_RT_WARNING_LEVEL>::get() > 0;
   AddToConfigHandle(
       ::jit_compiler::option::JITEnableVerbose::set(DebugEnabled));
-
-  auto SetUpOption = [](std::string Value) {
-    ::jit_compiler::JITEnvVar Option(Value.size());
-    memcpy(static_cast<void *>(Option.begin()),
-           static_cast<const void *>(Value.data()), Value.size());
+  auto SetUpOption = [](const std::string &Value) {
+    ::jit_compiler::JITEnvVar Option(Value.size(), Value.data());
```
Comment (victor-eds) [inline sycl/source/detail/jit_compiler.cpp:667]:
```suggestion
    ::jit_compiler::JITEnvVar Option(Value.begin(), Value.end());
```
Should work with above.

File: sycl-jit/common/include/DynArray.h (hunk 6 above)
Comment (jchlanda) [reply sycl-jit/common/include/DynArray.h]: Sure, if you prefer iterator approach, I'm cool with that. Done.

File: sycl/source/detail/jit_compiler.cpp (hunk 7 above)
Comment (jchlanda) [reply sycl/source/detail/jit_compiler.cpp]: Yeap, done now.

**Review by sommerlukas — APPROVED** [review]

**Pull request merged**